### PR TITLE
feat!(lsp): custom callback for semantic token highlighting

### PR DESCRIPTION
**Pull request by jdrouhard**
This is simply a draft that adds support for allowing a user to provide a callback to add the actual highlight or extmark if desired. The discussion about allowing more customizability for semantic tokens is discussed in #21576.

I'm absolutely open to suggestions or alternate takes on this. My original proposal in the linked issue was to extend the highlight group infrastructure to allow adding N modifiers to a main highlight group name with a `#` separator, but as discussed, even implementing this may not cover all possible use cases someone could want or come up with. The most flexible solution seemed to be to allow users to receive the semantic tokens in a callback so they can perform their own logic and decide how to add the highlights.

This PR also changes the _default_ for semantic token highlighting to be _not_ enabled. Users won't have to erase the `semanticTokensProvider` from their client capabilities to opt-out any longer, but those people who do still want it will need to add a `start(...)` call to their `LspAttach`/`on_attach()` callbacks to opt-in.

Doing it this way has some advantages:

1. The sky's the limit. Savvy/advanced users can do whatever they wish with the semantic tokens as they are displayed on the screen.
2. We don't lock ourselves into a highlighting API that ends up not being flexible enough for the things people can come up with.
3. It doesn't preclude us from adding something more railroaded/"official" down the line (like the original `#` proposal), as long as we always leave this "escape hatch" in place.
4. It gives those who want the added flexibility something they can work with now, instead of waiting for that more baked-in API.

Disadvantages:
1. There is still no "de-facto" way for colorscheme authors to stylize specific combinations of tokens and modifiers. This can likely be "fixed" or worked around by colorschemes or plugins that provide their own callbacks that add highlights/extmarks with specific combinations in specific ways.
2. Adding a callback for tokens locks us a little since the token format itself must not change without breaking the API.

There are two ways this callback could work. The way I have it in this PR is the most flexible, but an alternative could be that the callback should _return_ a highlight group name (or names in a list, or names and priorities in a list) and the built-in semantic token engine would continue to apply the extmark using either the callback's return value, or what it's doing now if there isn't a callback (the token type prepended with `@` and another extmark per modifier).

As an example, I have this in my `LspAttach` autocmd callback:

```lua
vim.lsp.semantic_tokens.start(bufnr, client.id, {
  hl_cb = function(bufnr, ns, token)
    local hl_group = '@' .. token.type
    if token.type == 'parameter' and
        not vim.tbl_contains(token.modifiers, 'declaration') then
      hl_group = hl_group .. '.reference'
    end
    if token.type == 'variable' and vim.tbl_contains(token.modifiers, 'declaration') then
      hl_group = hl_group .. '.declaration'
    end
    api.nvim_buf_set_extmark(bufnr, ns, token.line, token.start_col, {
      hl_group = hl_group,
      end_col = token.end_col,
      priority = vim.highlight.priorities.semantic_tokens,
      strict = false,
    })
  end
})
```

And I have custom highlights for `@variable.declaration` and `@parameter.reference`.

Pinging @mfussenegger @clason @swarn @rebelot @lewis6991 and @justinmk for input.

### CODE REVIEW

**Comment by mfussenegger** [comment]
Not a fan of this as it couples color scheme with concrete semantic token highlighter implementations. That will make changing color-schemes a lot more troublesome and color-scheme authors need to learn one more thing and how to use it well.
I think customization of syntax highlighting should be the colorscheme. Not Lua logic that applies extmarks.

I also kinda see both TS and LSP as implementation details when it comes to syntax highlighting. Users and colorscheme authors shouldn't really need to know where/what associates a code fragment with a highlight group, but only have to know which highlight groups exist (and with what modifiers).

And I'm still in favour of the opt-in by default. Opting into LSP is already a choice and Neovim already made the choice that syntax highlighting should be enabled by default. Telling users to copy 20 lines into their dotfiles would imho be a worse experience.

**Comment by swarn** [comment]
Since @jdrouhard has delivered exactly what I hoped for, it's not surprising that I think this is great.

Directly applying the highlight in the callback with `set_extmark` is smart. This puts all functionality in the hands of a user. This is now a "do *anything* with semantic tokens" callback that can be used for highlighting. Also, this can be used as the basis of future changes, right? That is, any highlighting system (combination of naming, API, and convention) that you care to design should be expressible with this, which is a good sign. For example, here's the current system:
```lua
vim.lsp.semantic_tokens.start(bufnr, client.id, {
  hl_cb = function(bufnr, ns, token)
    local ft_ext = "." .. vim.fn.getbufvar(bufnr, "&filetype")
    local hl_group = '@' .. token.type .. ft_ext

    api.nvim_buf_set_extmark(bufnr, ns, token.line, token.start_col, {
      hl_group = hl_group,
      end_col = token.end_col,
      priority = vim.highlight.priorities.semantic_tokens,
      strict = false,
    })
  end
})
```
...with a little change on my part; `class` tokens now highlight with `@class.{filetype}` instead of `@class`. Viva customization!

Is it possible for this callback to be used by default, unless replaced or disabled? Then you get nice default highlighting based on treesitter groups that colorschemes are already using.

I agree that writing callbacks for the LSP config that set extmarks — whew! — is not the easiest approach for someone who just wants one or two modifier-based highights, and should not be the last word on highlighting-using-tokens. But — trying not to repeat all of #21576, here — I believe:

1. The first step should be primarily concerned with the expressiveness and extensibility of highlighting-using-tokens. Maybe I'm wrong about that. I just watched "State of Neovim 2022" video from @justinmk; is it fair to say we are trying to find a balance between the first two goals of neovim, "extensible" and "usable?" In any case, this PR **is extensible**.

2. I feel like the "copy this code into your dotfiles" approach is exactly what has worked so far for LSP. Then with a little experience, people do great work to [package up sane defaults](https://github.com/VonHeikemen/lsp-zero.nvim), but we can still use the lower-level APIs to customize.

@mfussenegger, when you say "I think customization of syntax highlighting should be the colorscheme," I'm assuming you mean "expressed within `:hi` commands." Am I right? If so, I'm not against that, I just have the concerns listed in #21576; it's an API, that must be expressible within group names, that must be maintained going forward, and that will be very limited unless someone puts a lot of thought and work into it.

I'm not sure why this PR makes switching color schemes more difficult. If your callback applies the common highlight groups, it's good to go. If you define custom highlights, like `@class.myspecialhighlight`, the treesitter fallback mechanism has you covered. If you want to pick a color for your custom highlight in every color scheme you switch to, then you need to do [the same tricks](https://gist.github.com/romainl/379904f91fa40533175dfaec4c833f2f) you always have.

**Comment by rebelot** [comment]
> Not a fan of this as it couples color scheme with concrete semantic token highlighter implementations. That will make changing color-schemes a lot more troublesome and color-scheme authors need to learn one more thing and how to use it well. I think customization of syntax highlighting should be the colorscheme. Not Lua logic that applies extmarks.

I agree, but if this is the standpoint, and the only way for colorschemes to convey user preferences should be ultimately via set_hl, then either the hlgroup name or some of its opts MUST be able to convey that preferences.

As of now, to avoid the need of specifying every possible hlgroup, the dot-fallback DSL was implemented, and we already discussed how this makes sense only if the information associated to a text region is hierarchical. Perhaps extending the DSL for non-hierarchical fields may be an option? (as `#` was proposed) (e.g. requested `#aaa#bbb` matches defined `#aaa#bbb`, `#bbb#aaa`, `#aaa` and `#bbb`, longest number of matched `#-fields` wins [footnote](###footnote))

>I also kinda see both TS and LSP as implementation details when it comes to syntax highlighting. Users and colorscheme authors shouldn't really need to know where/what associates a code fragment with a highlight group, but only have to know which highlight groups exist (and with what modifiers).

I think that having the knowledge that `@variable.builtin` exists for TS, but that `@variable` is also **(incidentally!!)** a semantic token type and that `defaultLibrary` is usually the modifier associated to builtins is unavoidable for anyone who's trying to put some effort and consistency into their colorscheme. This is true whatever syntax is used for tokens/modifiers, # syntax, multiple extmarks, whatever, one will always think _"what the hell is the difference between builtin and defaultLibrary and why the latter is not working in my c++ file?"_. Perhaps it is advisable that the audience is provided some pointers about these two overlapping and complementary highlighters.

BTW, If someone wants to be concise, the proposed syntax `@variable#builtin#defaultLibrary` would match hl names provided both by TS and LSP tokens in one go. I'd even go the extra mile and add an additional and optional `:filetype` field at the end of the hlgroup name.

### tl;dr
My preferred solution would be a highlight spec DSL, it feels vim-y and expressive, but for the time being I think the callback approach of this PR would provide an instant boost to semantic highlights usability.

### footnote
[1]: in case only `#aaa` and `#bbb` are defined, the "problem" is how to determine a priority between the two

**Comment by mfussenegger** [comment]
> is it fair to say we are trying to find a balance between the first two goals of neovim, "extensible" and "usable?" In any case, this PR is extensible.

Extensibility is a goal, but the _what_, _why_ and _how_ matters.
Currently one could already implement any custom highlighter as plugin. (An example: https://github.com/icedman/nvim-textmate)


> I feel like the "copy this code into your dotfiles" approach is exactly what has worked so far for LSP. Then with a little experience, people do great work to package up sane defaults, but we can still use the lower-level APIs to customize.

It's also a major point of criticism. Some of which is very valid. Providing sane defaults _and_ providing the ability to customize is not necessarily contradicting. Figuring out a path to have _both_ is the challenge. Just because we can defer to user customization doesn't mean we should always resort to it. After all usability and strong defaults is _also_ a goal of Neovim.


> when you say "I think customization of syntax highlighting should be the colorscheme," I'm assuming you mean "expressed within :hi commands."

Kinda.

How I see it is that there is currently a split between:

1. A highlighter: A system that defines what parts of a buffer map to which highlight groups. There's currently at least 3 implementations (Syntax engine, Tree-sitter, LSP)
2. A colorscheme: Defines how highlight group map to colors and attributes like bold, italic, etc.

I think this split is overall a very thing to have and that we should try to retain it. If we add an extension point like proposed in this PR, the line between the two would get blurrier.


>  I just have the concerns listed in #21576; it's an API, that must be expressible within group names, that must be maintained going forward, and that will be very limited unless someone puts a lot of thought and work into it.

Yes, I think we ought to put that thought and work into it.


> I'm not sure why this PR makes switching color schemes more difficult. If your callback applies the common highlight groups, it's good to go. If you define custom highlights, like @class.myspecialhighlight, the treesitter fallback mechanism has you covered. If you want to pick a color for your custom highlight in every color scheme you switch to, then you need to do the same tricks you always have.

If a color scheme is created with a specific semantic token type/modifier mapping in mind, it won't work if the user has it customized in a different way. If two color scheme authors disagree on how semantic token type/modifier map to highlight groups you'd have to change the callback implementation when you switch color scheme. That's possible to do, but it's an extra burden.

---

> I think the callback approach of this PR would provide an instant boost to semantic highlights usability.

That may be right, but short term instant boosts have to be weighted against long term maintenance.

---


I think what might help to move forward would be concrete examples of how people would map semantic token types + modifiers to concrete highlight groups. Based on that we can figure out if it's possible to encode the majority of the cases into a string scheme.

**Comment by swarn** [comment]
I have to admit that when I see a working solution that already does what I want, I am excited to use it, and maybe not concerned enough with less immediate concerns like maintainability. But of course @mfussenegger is correct that maintainability is important.

When I look at this PR, I see few changes to the code. One callback option is added to one API function `vim.lsp.semantic_tokens.start()`, which is passed through unchanged to `STHighlighter:attach()`, and called once in `STHighlighter:on_win()`. Factoring out the "apply highlighting to a token" logic into its own function is good in any case, so the only issue is the API change.

I don't see that as a huge maintenance burden, but I have zero commits to this repo, so I should leave that judgement to people with more experience.

***

> Currently one could already implement any custom highlighter as plugin.

That is a good point! I would like to do that right now, because I am (obviously) excited about semantic tokens. My first thought is that I want to write custom handler functions, so that I can do something like:
``` lua
require('lspconfig').clangd.setup {
  handlers = {
    ["textDocument/semanticTokens/range"] = mymodule.fun1,
    ["textDocument/semanticTokens/full"] = mymodule.fun2,
    ["textDocument/semanticTokens/full/delta"] = mymodule.fun3,
    ["workspace/semanticTokens/refresh"] = mymodule.fun4,
  }
}
```
It seems like I will neeed to copy/re-implement all of `semantic_tokens.lua`. Many utilties from that, like `tokens_to_ranges`, are required. Basically, I just take this PR, copy it as a custom handler, and go from there.

If anyone has any thoughts/experience with that, I appreciate any comments! I have limited experience with the native LSP right now.

***

Would it make sense to customize the semantic highlights in the same way as other handlers? Something like:
```lua
vim.lsp.handlers["textDocument/semanticTokens"] = vim.lsp.with(
  vim.lsp.semantic_tokens.handler_function {
    hl_cb = my_hl_cb
  }
)
```

**Comment by jdrouhard** [comment]
Closing in favor of #22022
